sup3r/utilities/pytest/helpers.py

```diff
@@ -0,0 +1,405 @@
+        )
+    }
+    nc = xr.Dataset(coords=coords, data_vars=data_vars)
+    nc.to_netcdf(outfile)
+
+
+def make_fake_dset(shape, features, const=None):
+    """Make dummy data for tests."""
+
+    lats = np.linspace(70, -70, shape[0])
+    lons = np.linspace(-150, 150, shape[1])
+    lons, lats = np.meshgrid(lons, lats)
+    time = pd.date_range('2023-01-01', '2023-12-31', freq='60min')[: shape[2]]
+    dims = (
+        'time',
+        Dimension.PRESSURE_LEVEL,
+        Dimension.SOUTH_NORTH,
+        Dimension.WEST_EAST,
+    )
+    coords = {}
+
+    if len(shape) == 4:
+        levels = np.linspace(1000, 0, shape[3])
+        coords[Dimension.PRESSURE_LEVEL] = levels
+    coords['time'] = time
+    coords[Dimension.LATITUDE] = (
+        (Dimension.SOUTH_NORTH, Dimension.WEST_EAST),
+        lats,
+    )
+    coords[Dimension.LONGITUDE] = (
+        (Dimension.SOUTH_NORTH, Dimension.WEST_EAST),
+        lons,
+    )
+
+    dims = (
+        'time',
+        Dimension.PRESSURE_LEVEL,
+        Dimension.SOUTH_NORTH,
+        Dimension.WEST_EAST,
+    )
+    trans_axes = (2, 3, 0, 1)
+    if len(shape) == 3:
+        dims = ('time', *dims[2:])
+        trans_axes = (2, 0, 1)
+    data_vars = {}
+    for f in features:
+        if 'zg' in f:
+            data = da.random.uniform(10, 100, shape)
+        elif 'orog' in f:
+            data = da.random.uniform(0, 1, shape)
+        else:
+            data = da.random.uniform(-1, 1, shape)
+        data_vars[f] = (
+            dims[: len(shape)],
+            da.transpose(
+                np.full(shape, const) if const is not None else data,
+                axes=trans_axes,
+            ),
+        )
+    nc = xr.Dataset(coords=coords, data_vars=data_vars)
+    return nc.astype(np.float32)
+
+
+def make_fake_nc_file(file_name, shape, features):
+    """Make nc file with dummy data for tests."""
+    nc = make_fake_dset(shape, features)
+    nc.to_netcdf(file_name)
+
+
+class DummyData(Container):
+    """Dummy container with random data."""
+
+    def __init__(self, data_shape, features):
+        super().__init__()
+        self.data = make_fake_dset(data_shape, features)
+
+
+class DummySampler(Sampler):
+    """Dummy container with random data."""
+
+    def __init__(
+        self, sample_shape, data_shape, features, batch_size, feature_sets=None
+    ):
+        data = make_fake_dset(data_shape, features=features)
+        super().__init__(
+            Sup3rDataset(high_res=data),
+            sample_shape,
+            batch_size=batch_size,
+            feature_sets=feature_sets,
+        )
+
+
+def test_sampler_factory(SamplerClass):
+    """Build test samplers which track indices."""
+
+    class SamplerTester(SamplerClass):
+        """Keep a record of sample indices for testing."""
+
+        def __init__(self, *args, **kwargs):
+            self.index_record = []
+            super().__init__(*args, **kwargs)
+
+        def get_sample_index(self, **kwargs):
```

Comment:
Can we consider just adding this into the default batch handler infrastructure? Can be nice to know the historical sample indices and will simplify things a bit. Shouldn't be too much compute? We could always just store the index record for only the most recent sample.

Reply:
I didn't want to do this bc itll likely go unused and mostly only helps with testing. We added some stuff last time to make testing easier and it resulted in clutter.

Reply:
Just removing the testing code. I like having some way to pull the observation indices when verifying my training setup. I can use the test handlers but seems like a lightweight feature that could be easily included in the base handlers. I don't feel strongly about this so no worries either way, it's just a nice feature.

Reply:
For now I think Ill leave it. If you find that you are constantly having to use the test handlers lets add it in.